pennylane/decomposition/decomposition_graph.py

```diff
@@ -0,0 +1,243 @@
+        are conveniently wrapped in a ``CompressedResourceOp``.
+
+        """
+
+        if op_node in self._op_node_indices:
+            return self._op_node_indices[op_node]
+
+        op_node_idx = self._graph.add_node(op_node)
+        self._op_node_indices[op_node] = op_node_idx
+        if op_node.op_type.__name__ in self._target_gate_set:
+            self._target_gate_indices.add(op_node_idx)
+            return op_node_idx
+
+        for decomposition in self._get_decompositions(op_node.op_type):
+            d_node_idx = self._recursively_add_decomposition_node(decomposition, op_node.params)
+            self._graph.add_edge(d_node_idx, op_node_idx, 0)
+
+        return op_node_idx
+
+    def _recursively_add_decomposition_node(self, rule: DecompositionRule, params: dict) -> int:
+        """Recursively adds a decomposition node to the graph.
+
+        A decomposition node is defined by a decomposition rule and a first-order resource estimate
+        of this decomposition as computed with resource params passed from the operator node.
+
+        """
+
+        resource_decomp = rule.compute_resources(**params)
+        d_node = _DecompositionNode(rule, resource_decomp)
+        d_node_idx = self._graph.add_node(d_node)
+        all_ops = [op for op, count in resource_decomp.gate_counts.items() if count > 0]
+        for op in all_ops:
+            op_node_idx = self._recursively_add_op_node(op)
+            self._graph.add_edge(op_node_idx, d_node_idx, (op_node_idx, d_node_idx))
+        return d_node_idx
+
+    def solve(self, lazy=True):
+        """Solves the graph using the dijkstra search algorithm.
+
+        Args:
+            lazy (bool): If True, the dijkstra search will stop once optimal decompositions are
+                found for all operations that we originally want to decompose. Otherwise, the
+                entire graph will be explored.
+
+        """
+        self._visitor = _DecompositionSearchVisitor(self._graph, self._original_ops_indices, lazy)
+        start = self._graph.add_node("dummy")
+        self._graph.add_edges_from(
+            [(start, op_node_idx, 1) for op_node_idx in self._target_gate_indices]
+        )
+        rx.dijkstra_search(
+            self._graph,
+            source=[start],
+            weight_fn=self._visitor.edge_weight,
+            visitor=self._visitor,
+        )
+        self._graph.remove_node(start)
+        if self._visitor.unsolved_op_indices:
+            unsolved_ops = [self._graph[op_idx] for op_idx in self._visitor.unsolved_op_indices]
+            op_names = set(op.op_type.__name__ for op in unsolved_ops)
+            raise DecompositionError(
+                f"Decomposition not found for {op_names} to the gate set {self._target_gate_set}"
+            )
+
+    def resource_estimates(self, op) -> Resources:
+        """Returns the resource estimates for a given operator.
```

Comment:
In my head, resources are plural and there is a single estimate returned by this function. Not sure whether there is a consensus about this aspect. It's really a minor thing, but might be nice to ensure unification across code base :)

Reply:
I guess `resource_estimate` would make more sense.

Reply:
Just changed it to `resource_estimate`.

Reply:
Sounds good, thanks!